File: C++_programming/SinglyLL.cpp
#include<iostream>
using namespace std;

struct  node
{
    int data;
    struct node * next;
};

typedef struct node NODE;
typedef struct node * PNODE;
typedef struct node ** PPNODE;

class SinglyLL
{
    public:
        int iCount;
        PNODE first;

        SinglyLL()
        {
            cout<<"Object of SinglyLL gets created \n";
            first = NULL;
            iCount = 0;
        }

        void InsertFirst(int no)
        {
            PNODE newn = NULL;
            newn = new NODE;

            newn ->data = no;
            newn -> next = NULL;

            if(first == NULL)
            {
                first = newn;
            }
            else
            {
                newn -> next = first ;
                first = newn;
            }
            iCount ++;
        }

        void InsertLast(int no)
        {
            PNODE newn = NULL;
            PNODE temp = NULL;
            newn = new NODE;

            newn ->data = no;
            newn-> next = NULL;
            
            if(first == NULL)
            {
                first = newn;
            }
            else
            {
                temp = first;
                while(temp->next != NULL)
                {
                    temp = temp -> next;
                }
                    temp ->next = newn;

            }

            iCount++;
            
        }

        void InsertAtPos(int no , int pos)
        {
            PNODE temp = NULL;
            PNODE newn = NULL;

            newn = new NODE;
            newn->data = no;
            newn->next = NULL;
            int iSize = 0 ;
            int iCnt =0;
            iSize =Count();

            if(pos ==1)
            {
                InsertFirst(no);
            }
            else if(pos == iSize + 1)
            {
                InsertLast(no);
            }
            else
            {
               newn = new NODE;
               
               newn->data = no;
               newn->next = NULL;
               temp =(first);

               for(iCnt =1 ; iCnt < pos -1 ; iCnt++)
               {
                    temp = temp->next;
               }

                newn ->next = temp ->next;
                temp ->next = newn;

            }
            iCount ++;
        }

        void DeleteFirst()
        {
            PNODE temp = NULL;
            if(first == NULL)
            {
                return;
            }
            else if( first->next == NULL)
            {
                free(first);
                first ->next =NULL;
            }
            else
            {
                temp = first;
                first = first ->next;
                free(temp);
                
            }
            iCount--;
        }

        void DeleteLast()
        {
             PNODE temp = NULL;
            if(first == NULL)
            {
                return;
            }
            else if( first->next == NULL)
            {
                free(first);
                first ->next =NULL;
            }
            else
            { 
                temp = first;
               while(temp ->next ->next != NULL)
               {
                    temp = temp ->next;
               }
                    free(temp->next);
                    temp ->next =NULL;
                
            }
            iCount--;
        }

        void DeleteAtpos(int pos)
        {
           PNODE temp = NULL;
           PNODE target = NULL;

           int iCnt = 0, iSize = 0;
           iSize = Count();

           if(pos == 1)
           {
            DeleteFirst();
           }
           else if(pos == iSize + 1)
           {
            DeleteLast();
           }
           else
           {
                temp = first;

                for(iCnt = 1; iCnt < pos -1 ; iCnt++)
                {
                    temp = temp ->next;
                }
                target = temp -> next;
                temp ->next = target ->next;
                free(target);

           }
           iCount--;
        }

        void Display()
        {
            PNODE temp = NULL;

            temp = first;
            while(temp != NULL)
            {
                cout<<"|"<<temp->data<<"|->";
                temp = temp -> next;
            }
            cout<<"NULL \n";
        }

        int Count()
        {
            return iCount;
        }
 

};

int main ()
{
    SinglyLL sobj;
    int iRet = 0;

    sobj.InsertFirst(51);
    sobj.InsertFirst(21);
    sobj.InsertFirst(11);

    sobj.Display();
    iRet =sobj.Count();
    cout<<"number of node are :"<<iRet<<"\n";

    sobj.InsertLast(151);
    sobj.InsertLast(121);
    sobj.InsertLast(111);

    sobj.Display();
    iRet = sobj.Count();
    cout<<"Number of node are :"<<iRet<<"\n";

    sobj.InsertAtPos(33,3);
    sobj.Display();
    iRet = sobj.Count();
    cout<<"Number of node are:"<<iRet<<"\n";

    sobj.DeleteFirst();
    sobj.Display();
    iRet = sobj.Count();
    cout<<"Number of node are:"<<iRet<<"\n";

    sobj.DeleteLast();
    sobj.Display();
    iRet = sobj.Count();
    cout<<"Number of node are:"<<iRet<<"\n";

    sobj.DeleteAtpos(3);
    sobj.Display();
    iRet = sobj.Count();
    cout<<"Number of node are:"<<iRet<<"\n";



    return 0;
}

File: python/singlyLL.py

class node:
    def __init__(self,value):
        self.Data = value
        self.next = None

class SinglyLL:
    
    def __init__(self):
        self.first = None
        self.iCount = 0
    
    
    def InsertFirst(self,no):
        newn = node(no)

      
        if(self.first == None):
            self.first = newn
      
        else:
            newn.next= self.first
            self.first = newn
        
        self.iCount = self.iCount + 1

    
    def InsertLast(self,no):
        newn = node(no)

        
        if(self.first == None):
            self.first = newn
        
        else:
            temp = self.first

            while(temp.next != None):
                temp = temp.next
            temp.next = newn
        
        self.iCount = self.iCount + 1

    
    def InsertAtPos(self,no,pos):
        
        if(pos < 1 or pos > (self.iCount+1)):
            print("Invalid Position ")
            return
        if(pos == 1):
            self.InsertFirst(no)
            return
        elif(pos == self.iCount+1):
            self.InsertLast(no)
            return
        else:
            newn = node(no)
            temp = self.first

            for i in range(1,pos-1):
                temp = temp.next

            newn.next = temp.next
            temp.next = newn
            
            self.iCount = self.iCount + 1
            
        


  
    def DeleteFirst(self):
        if(self.first == None):
            return
        
        temp = self.first
        self.first = self.first.next

        del temp 

        self.iCount = self.iCount -1
    
    def DeleteLast(self):
        
        if(self.first == None):
            return
      
        if(self.first.next == None):
            del self.first
            self.first = None
            self.iCount = 0
    
        else:
            temp = self.first

            while(temp.next.next != None):
                temp = temp.next
        
            del temp.next        
            temp.next= None
            self.iCount = self.iCount -1

    def DeleteAtPos(self,pos):
        
        if(pos < 1 or pos > (self.iCount)):
            print("Invalid Position ")
            return
        if(pos == 1):
            self.DeleteFirst()
            return
        elif(pos == self.iCount):
            self.DeleteLast()
            return
        else:
            temp = self.first

            for i in range(1,pos-1):
                temp = temp.next

            temp.next = temp.next.next
            
            self.iCount = self.iCount - 1
            
        
    
    
    def Count(self):
        return self.iCount 
  
    def Display(self):
        temp = self.first

        
        while temp is not None:
            print("| ",temp.Data,"|->",end=" ")
            temp = temp.next
        print(None)
        
def main():
    sobj = SinglyLL()

    sobj.InsertFirst(101) 
    sobj.InsertFirst(51)
    sobj.InsertFirst(21)
    sobj.InsertFirst(11)

    print("elements of LinkedList are :")
    sobj.Display()
    print("Number of elments in linked list are ",sobj.Count())

    sobj.InsertLast(111)
    sobj.InsertLast(121)

    print("elements of LinkedList are :")
    sobj.Display()
    print("Number of elments in linked list are ",sobj.Count())


    sobj.InsertAtPos(75,4)
    print("elements of LinkedList are :")
    sobj.Display()
    print("Number of elments in linked list are ",sobj.Count())


    sobj.DeleteFirst()
    print("elements of LinkedList are :")
    sobj.Display()
    print("Number of elments in linked list are ",sobj.Count())

     
    sobj.DeleteLast()
    print("elements of LinkedList are :")
    sobj.Display()
    print("Number of elments in linked list are ",sobj.Count())

    sobj.DeleteAtPos(3)
    print("elements of LinkedList are :")
    sobj.Display()
    print("Number of elments in linked list are ",sobj.Count())

if __name__ == "__main__":
    main()

